File: src/models/quest.py
from sqlalchemy import BigInteger, Column, ForeignKey, Integer, Text, UniqueConstraint, Index
from sqlalchemy.orm import Mapped, mapped_column, relationship
# from sqlalchemy.dialects.postgresql import JSONB # Removed
from sqlalchemy import Enum as SQLAlchemyEnum, Boolean, CheckConstraint # Added CheckConstraint
import datetime # Added for PlayerQuestProgress
from typing import TYPE_CHECKING, Optional, Dict, Any, List

from .base import Base, TimestampMixin # Added TimestampMixin
from .enums import RelationshipEntityType, QuestStatus # Assuming QuestStatus is in enums
from .custom_types import JsonBForSQLite # Added

if TYPE_CHECKING:
    from .guild import GuildConfig
    from .player import Player # For PlayerQuestProgress relationship
    from .party import Party # Added for PlayerQuestProgress relationship
    from .quest import QuestStep # For GeneratedQuest relationship to current_step and PlayerQuestProgress

class Questline(Base, TimestampMixin): # Added TimestampMixin
    __tablename__ = "questlines"
    id: Mapped[int] = mapped_column(Integer, primary_key=True, index=True)
    guild_id: Mapped[int] = mapped_column(BigInteger, ForeignKey("guild_configs.id", ondelete="CASCADE"), nullable=False, index=True)
    static_id: Mapped[str] = mapped_column(Text, nullable=False, index=True)
    title_i18n: Mapped[Dict[str, str]] = mapped_column(JsonBForSQLite, nullable=False, default=lambda: {}) # Renamed from name_i18n
    description_i18n: Mapped[Dict[str, str]] = mapped_column(JsonBForSQLite, nullable=False, default=lambda: {}) # Changed

    starting_quest_static_id: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    is_main_storyline: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)
    required_previous_questline_static_id: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    properties_json: Mapped[Optional[Dict[str, Any]]] = mapped_column(JsonBForSQLite, nullable=True, default=lambda: {})

    quests: Mapped[List["GeneratedQuest"]] = relationship(back_populates="questline")
    __table_args__ = (UniqueConstraint("guild_id", "static_id", name="uq_questline_guild_static_id"),)
    def __repr__(self) -> str:
        return f"<Questline(id={self.id}, static_id='{self.static_id}', guild_id={self.guild_id})>"

class GeneratedQuest(Base, TimestampMixin): # Added TimestampMixin
    __tablename__ = "generated_quests"
    id: Mapped[int] = mapped_column(Integer, primary_key=True, index=True)
    guild_id: Mapped[int] = mapped_column(BigInteger, ForeignKey("guild_configs.id", ondelete="CASCADE"), nullable=False, index=True)
    static_id: Mapped[str] = mapped_column(Text, nullable=False, index=True) # Unique within guild
    title_i18n: Mapped[Dict[str, str]] = mapped_column(JsonBForSQLite, nullable=False, default=lambda: {}) # Changed
    description_i18n: Mapped[Dict[str, str]] = mapped_column(JsonBForSQLite, nullable=False, default=lambda: {}) # Changed
    questline_id: Mapped[Optional[int]] = mapped_column(Integer, ForeignKey("questlines.id", ondelete="SET NULL"), nullable=True, index=True)
    questline: Mapped[Optional["Questline"]] = relationship(back_populates="quests") # Optional
    giver_entity_type: Mapped[Optional[RelationshipEntityType]] = mapped_column(SQLAlchemyEnum(RelationshipEntityType, name="relationship_entity_type_enum", create_type=False), nullable=True)
    giver_entity_id: Mapped[Optional[int]] = mapped_column(BigInteger, nullable=True)
    min_level: Mapped[Optional[int]] = mapped_column(Integer, nullable=True) # Renamed from required_level
    is_repeatable: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)
    rewards_json: Mapped[Optional[Dict[str, Any]]] = mapped_column(JsonBForSQLite, nullable=True, default=lambda: {}) # Changed
    properties_json: Mapped[Optional[Dict[str, Any]]] = mapped_column(JsonBForSQLite, nullable=True, default=lambda: {}) # Added (ai_metadata_json can be part of this or separate if needed)
    ai_metadata_json: Mapped[Optional[Dict[str, Any]]] = mapped_column(JsonBForSQLite, nullable=True, default=lambda: {}) # Changed
    steps: Mapped[List["QuestStep"]] = relationship(back_populates="quest", order_by="QuestStep.step_order", cascade="all, delete-orphan")
    player_progress: Mapped[List["PlayerQuestProgress"]] = relationship(back_populates="quest", cascade="all, delete-orphan")
    __table_args__ = (
        UniqueConstraint("guild_id", "static_id", name="uq_generated_quest_guild_static_id"),
        Index('ix_generated_quests_giver', 'guild_id', 'giver_entity_type', 'giver_entity_id'),
    )
    def __repr__(self) -> str:
        return f"<GeneratedQuest(id={self.id}, static_id='{self.static_id}', guild_id={self.guild_id})>"

class QuestStep(Base, TimestampMixin): # Added TimestampMixin
    __tablename__ = "quest_steps"
    id: Mapped[int] = mapped_column(Integer, primary_key=True, index=True)
    quest_id: Mapped[int] = mapped_column(Integer, ForeignKey("generated_quests.id", ondelete="CASCADE"), nullable=False, index=True)
    quest: Mapped["GeneratedQuest"] = relationship(back_populates="steps") # Optional
    step_order: Mapped[int] = mapped_column(Integer, nullable=False, default=1)
    title_i18n: Mapped[Dict[str, str]] = mapped_column(JsonBForSQLite, nullable=False, default=lambda: {}) # Changed
    description_i18n: Mapped[Dict[str, str]] = mapped_column(JsonBForSQLite, nullable=False, default=lambda: {}) # Changed
    required_mechanics_json: Mapped[Optional[Dict[str, Any]]] = mapped_column(JsonBForSQLite, nullable=True, default=lambda: {}) # Changed
    abstract_goal_json: Mapped[Optional[Dict[str, Any]]] = mapped_column(JsonBForSQLite, nullable=True, default=lambda: {}) # Changed
    consequences_json: Mapped[Optional[Dict[str, Any]]] = mapped_column(JsonBForSQLite, nullable=True, default=lambda: {}) # Changed
    next_step_order: Mapped[Optional[int]] = mapped_column(Integer, nullable=True)
    properties_json: Mapped[Optional[Dict[str, Any]]] = mapped_column(JsonBForSQLite, nullable=True, default=lambda: {})
    __table_args__ = (UniqueConstraint("quest_id", "step_order", name="uq_quest_step_order"),)
    def __repr__(self) -> str:
        return f"<QuestStep(id={self.id}, quest_id={self.quest_id}, order={self.step_order})>"

class PlayerQuestProgress(Base, TimestampMixin): # Added TimestampMixin
    __tablename__ = "player_quest_progress"
    id: Mapped[int] = mapped_column(Integer, primary_key=True, index=True) # Own PK for this table
    guild_id: Mapped[int] = mapped_column(BigInteger, ForeignKey("guild_configs.id", ondelete="CASCADE"), nullable=False, index=True)
    player_id: Mapped[Optional[int]] = mapped_column(Integer, ForeignKey("players.id", ondelete="CASCADE"), nullable=True, index=True) # Made nullable
    player: Mapped[Optional["Player"]] = relationship(back_populates="quest_progress")
    party_id: Mapped[Optional[int]] = mapped_column(Integer, ForeignKey("parties.id", ondelete="CASCADE"), nullable=True, index=True) # Added
    party: Mapped[Optional["Party"]] = relationship() # Added, no back_populates needed if one-way or handled elsewhere

    quest_id: Mapped[int] = mapped_column(Integer, ForeignKey("generated_quests.id", ondelete="CASCADE"), nullable=False, index=True)
    quest: Mapped["GeneratedQuest"] = relationship(back_populates="player_progress")
    current_step_id: Mapped[Optional[int]] = mapped_column(Integer, ForeignKey("quest_steps.id", ondelete="SET NULL"), nullable=True, index=True)
    current_step: Mapped[Optional["QuestStep"]] = relationship()
    status: Mapped[QuestStatus] = mapped_column(SQLAlchemyEnum(QuestStatus, name="quest_status_enum", create_type=False), nullable=False, default=QuestStatus.NOT_STARTED, index=True)
    progress_data_json: Mapped[Optional[Dict[str, Any]]] = mapped_column(JsonBForSQLite, nullable=True, default=lambda: {}) # Changed

    accepted_at: Mapped[Optional[datetime.datetime]] = mapped_column(JsonBForSQLite.DATETIME_TYPE, nullable=True) # Changed to DATETIME_TYPE for SQLite compatibility
    completed_at: Mapped[Optional[datetime.datetime]] = mapped_column(JsonBForSQLite.DATETIME_TYPE, nullable=True) # Changed to DATETIME_TYPE for SQLite compatibility

    __table_args__ = (
        UniqueConstraint("guild_id", "player_id", "quest_id", name="uq_player_quest_guild"),
        UniqueConstraint("guild_id", "party_id", "quest_id", name="uq_party_quest_guild"),
        CheckConstraint("player_id IS NOT NULL OR party_id IS NOT NULL", name="cc_player_or_party_id_not_null"),
        Index('ix_player_quest_progress_player_quest', 'player_id', 'quest_id'), # Explicit combined index
        Index('ix_player_quest_progress_party_quest', 'party_id', 'quest_id'),   # Explicit combined index
    )
    def __repr__(self) -> str:
        if self.player_id:
            return f"<PlayerQuestProgress(player_id={self.player_id}, quest_id={self.quest_id}, status='{self.status.value}')>"
        elif self.party_id:
            return f"<PlayerQuestProgress(party_id={self.party_id}, quest_id={self.quest_id}, status='{self.status.value}')>"
        return f"<PlayerQuestProgress(id={self.id}, quest_id={self.quest_id}, status='{self.status.value}')>"

File: src/ui/src/types/quest.ts
// src/ui/src/types/quest.ts

/**
 * Общий интерфейс для пагинированных ответов API.
 */
export interface PaginatedResponse<T> {
  items: T[];
  total: number;
  page: number;
  limit: number;
  total_pages: number;
}

/**
 * Локализованные строки (ключ - код языка, значение - текст).
 */
export type LocaleRecord = Record<string, string>;

/**
 * Произвольные JSON данные.
 */
export type JsonData = Record<string, any>;

/**
 * Данные для цепочки квестов (Questline).
 */
export interface QuestlineData {
  id: number;
  guild_id: string; // В API guild_id часто строка из-за больших чисел Discord ID
  static_id: string;
  title_i18n: LocaleRecord;
  description_i18n?: LocaleRecord | null;
  starting_quest_static_id?: string | null;
  is_main_storyline: boolean;
  required_previous_questline_static_id?: string | null;
  properties_json?: JsonData | null;
  created_at: string; // ISO datetime string
  updated_at: string; // ISO datetime string
}

/**
 * Пейлоад для создания/обновления Questline.
 * Для команды update используется field_to_update и new_value,
 * но UI может собирать полный объект и отправлять только измененные поля,
 * или использовать более гранулярный подход.
 * Этот Payload больше для создания.
 */
export interface QuestlinePayload {
  static_id: string;
  title_i18n_json: string; // JSON string of LocaleRecord
  description_i18n_json?: string; // JSON string of LocaleRecord
  starting_quest_static_id?: string;
  is_main_storyline?: boolean;
  prev_questline_id?: string; // static_id of previous questline
  properties_json?: string; // JSON string of JsonData
}

/**
 * Данные для генерируемого квеста (GeneratedQuest).
 */
export interface GeneratedQuestData {
  id: number;
  guild_id: string;
  static_id: string;
  title_i18n: LocaleRecord;
  description_i18n?: LocaleRecord | null;
  questline_id?: number | null;
  // giver_entity_type и giver_entity_id не управляются напрямую через CRUD UI по задаче
  // min_level также
  is_repeatable: boolean;
  rewards_json?: JsonData | null;
  properties_json?: JsonData | null; // Может содержать quest_type, difficulty и др.
  ai_metadata_json?: JsonData | null;
  created_at: string;
  updated_at: string;
  // UI может захотеть загружать шаги отдельно или вместе с квестом
  steps?: QuestStepData[];
}

/**
 * Пейлоад для создания/обновления GeneratedQuest.
 */
export interface GeneratedQuestPayload {
  static_id: string;
  title_i18n_json: string; // JSON string
  description_i18n_json?: string; // JSON string
  quest_type?: string; // Будет помещен в properties_json.quest_type
  questline_id?: number;
  is_repeatable?: boolean;
  properties_json?: string; // JSON string
  rewards_json?: string; // JSON string
}

/**
 * Данные для шага квеста (QuestStep).
 */
export interface QuestStepData {
  id: number;
  quest_id: number;
  step_order: number;
  title_i18n: LocaleRecord;
  description_i18n: LocaleRecord; // В модели not nullable, default {}
  required_mechanics_json?: JsonData | null;
  abstract_goal_json?: JsonData | null;
  consequences_json?: JsonData | null;
  next_step_order?: number | null;
  properties_json?: JsonData | null;
  created_at: string;
  updated_at: string;
}

/**
 * Пейлоад для создания/обновления QuestStep.
 */
export interface QuestStepPayload {
  quest_id: number;
  step_order: number;
  title_i18n_json: string; // JSON string
  description_i18n_json: string; // JSON string
  required_mechanics_json?: string; // JSON string
  abstract_goal_json?: string; // JSON string
  consequences_json?: string; // JSON string
  next_step_order?: number;
  properties_json?: string; // JSON string
}

/**
 * Статусы квеста (должны соответствовать Enum в бэкенде).
 */
export enum UIQuestStatus {
  NOT_STARTED = "NOT_STARTED",
  IN_PROGRESS = "IN_PROGRESS",
  COMPLETED = "COMPLETED",
  FAILED = "FAILED",
  CANCELLED = "CANCELLED", // Если есть такой статус
  ON_HOLD = "ON_HOLD" // Если есть
}

/**
 * Данные о прогрессе выполнения квеста игроком/партией (PlayerQuestProgress).
 */
export interface PlayerQuestProgressData {
  id: number;
  guild_id: string;
  player_id?: number | null;
  party_id?: number | null;
  quest_id: number;
  current_step_id?: number | null;
  status: UIQuestStatus; // Используем UIQuestStatus Enum
  progress_data_json?: JsonData | null;
  accepted_at?: string | null; // ISO datetime string
  completed_at?: string | null; // ISO datetime string
  created_at: string;
  updated_at: string;
}

/**
 * Пейлоад для создания PlayerQuestProgress.
 */
export interface PlayerQuestProgressPayload {
  quest_id: number;
  player_id?: number;
  party_id?: number;
  status?: UIQuestStatus;
  current_step_id?: number;
  progress_data_json?: string; // JSON string
  accepted_at_iso?: string; // ISO datetime string
}

/**
 * Пейлоад для обновления PlayerQuestProgress (попольное).
 * Используется в команде /master_quest progress_update
 */
export interface PlayerQuestProgressUpdatePayload {
  field_to_update: "status" | "current_step_id" | "progress_data_json";
  new_value: string; // Имя Enum для status, число для ID, JSON-строка для json, "None"
}
